src/cached_storage.rs

```diff
@@ -0,0 +1,398 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ------------------------------------------------------------------------------
+ */
+
+use std::collections::VecDeque;
+use std::io;
+use std::path::PathBuf;
+use std::sync::{Arc, RwLock, RwLockReadGuard, RwLockWriteGuard};
+
+use raft::{
+    self, eraftpb::{ConfState, Entry, HardState, Snapshot}, RaftState, Storage,
+};
+
+use fs_storage::FsStorage;
+use storage::StorageExt;
+
+pub const CACHE_SIZE: u64 = 10000;
+
+struct Cache {
+    entries: VecDeque<Entry>,
+    _cache_size: u64,
+}
+
+impl Cache {
+    fn with_cache_size(cache_size: u64) -> Cache {
+        Cache {
+            entries: VecDeque::new(),
+            _cache_size: cache_size,
+        }
+    }
+
+    fn first_index(&self) -> u64 {
+        match self.entries.front() {
+            Some(entry) => entry.index,
+            None => 1,
+        }
+    }
+
+    fn last_index(&self) -> u64 {
+        match self.entries.back() {
+            Some(entry) => entry.index,
+            None => 0,
+        }
+    }
+
+    fn entries(&self, low: u64, high: u64, _max_size: u64) -> Result<Vec<Entry>, raft::Error> {
+        if self.entries.is_empty() || low < self.first_index() {
+            return Err(raft::Error::Store(raft::StorageError::Compacted));
+        }
+
+        if high > self.last_index() + 1 {
+            return Err(raft::Error::Store(raft::StorageError::Unavailable));
+        }
+
+        Ok(self.entries
+            .iter()
+            .cloned()
+            .filter(|entry| entry.index >= low && entry.index < high)
+            .collect())
+    }
+
+    fn term(&self, idx: u64) -> Option<u64> {
+        self.entries
+            .iter()
+            .find(|ref entry| entry.index == idx)
+            .map(|entry| entry.term)
+    }
+
+    fn compact(&mut self, compact_index: u64) {
+        self.entries.retain(|entry| entry.index >= compact_index);
+    }
+
+    fn append(&mut self, entries: &[Entry]) {
+        entries
+            .into_iter()
+            .for_each(|entry| self.entries.push_back(entry.clone()));
+    }
+}
+
+pub struct CachedStorage<S: StorageExt> {
+    storage: S,
+    cache: Arc<RwLock<Cache>>,
+}
+
+impl CachedStorage<FsStorage> {
+    pub fn with_data_dir(data_dir: PathBuf) -> io::Result<Self> {
+        let storage = FsStorage::with_data_dir(data_dir).expect("Failed to create FsStorage");
+
+        Ok(CachedStorage {
+            storage,
+            cache: Arc::new(RwLock::new(Cache::with_cache_size(CACHE_SIZE))),
+        })
+    }
+}
```

Comment:
Instead of providing a constructor for a specific type of storage, the constructor should take an argument of a type that implements StorageExt. So you would have:
```
impl CachedStorage<S: StorageExt> {
  pub fn new(storage: S) -> Self {
        Ok(CachedStorage {
            storage,
            cache: Arc::new(RwLock::new(Cache::with_cache_size(CACHE_SIZE))),
        })
  }
}
```
and then later when you construct this you would have:
```
    CachedStorage::new(
        FsStorage::with_data_dir(
            get_path_config().data_dir).expect("Failed to create FsStorage"))
```